File: src/main/java/fr/lewon/bot/runner/util/BotPropertyParser.java
package fr.lewon.bot.runner.util;

import fr.lewon.bot.runner.bot.props.BotPropertyDescriptor;
import fr.lewon.bot.runner.bot.props.BotPropertyStore;
import fr.lewon.bot.runner.errors.InvalidBotPropertyValueException;
import fr.lewon.bot.runner.errors.MissingBotPropertyException;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Map;

@Component
public class BotPropertyParser {

    public BotPropertyStore parseParams(Map<String, String> params, List<BotPropertyDescriptor> targetBotPropertyDescriptors) throws InvalidBotPropertyValueException, MissingBotPropertyException {

        BotPropertyStore botPropertyStore = new BotPropertyStore();
        for (BotPropertyDescriptor propertyDescriptor : targetBotPropertyDescriptors) {
            if (!params.containsKey(propertyDescriptor.getKey()) && propertyDescriptor.isNeeded()) {
                throw new MissingBotPropertyException(propertyDescriptor);
            }
            botPropertyStore.addProperty(propertyDescriptor, this.parse(propertyDescriptor, params.get(propertyDescriptor.getKey())));
        }
        return botPropertyStore;
    }

    public Object parse(BotPropertyDescriptor descriptor, String value) throws InvalidBotPropertyValueException {
        if (value == null || value.isBlank()) {
            if (descriptor.isNullable()) {
                return null;
            }
            throw new InvalidBotPropertyValueException(value, descriptor);
        }

        Object parsedValue;
        try {
            parsedValue = descriptor.getType().parse(value);
        } catch (Exception e) {
            throw new InvalidBotPropertyValueException(value, descriptor, e);
        }

        if (descriptor.getAcceptedValues() == null || descriptor.getAcceptedValues().length == 0) {
            return parsedValue;
        }

        for (Object acceptedValue : descriptor.getAcceptedValues()) {
            if (acceptedValue.equals(parsedValue)) {
                return parsedValue;
            }
        }
        throw new InvalidBotPropertyValueException(value, descriptor);
    }

}

File: src/main/kotlin/fr/lewon/bot/runner/util/BotPropertyParser.kt
package fr.lewon.bot.runner.util

import fr.lewon.bot.runner.bot.props.BotPropertyDescriptor
import fr.lewon.bot.runner.bot.props.BotPropertyStore
import fr.lewon.bot.runner.errors.InvalidBotPropertyValueException
import fr.lewon.bot.runner.errors.MissingBotPropertyException

object BotPropertyParser {

    /**
     * Parses the parameters in the passed Map using the passed BotPropertyDescriptors, generating a BotPropertyStore. <br>
     *
     * @param params
     * @param targetBotPropertyDescriptors
     * @throws InvalidBotPropertyValueException
     * @throws MissingBotPropertyException
     */
    fun parseParams(
        params: Map<String, String?>,
        targetBotPropertyDescriptors: List<BotPropertyDescriptor>
    ): BotPropertyStore {
        val botPropertyStore = BotPropertyStore()
        for (propertyDescriptor in targetBotPropertyDescriptors) {
            if (params.containsKey(propertyDescriptor.key)) {
                botPropertyStore[propertyDescriptor] = this.parse(propertyDescriptor, params[propertyDescriptor.key])
            } else if (!propertyDescriptor.isNeeded) {
                botPropertyStore[propertyDescriptor] = propertyDescriptor.defaultValue
            } else {
                throw MissingBotPropertyException(propertyDescriptor)
            }
        }
        return botPropertyStore
    }

    @Suppress("UNCHECKED_CAST")
    private fun <T> parse(descriptor: BotPropertyDescriptor, value: String?): T? {
        if (value.isNullOrBlank()) {
            if (descriptor.isNullable) {
                return null
            }
            if (!descriptor.isNeeded) {
                return descriptor.defaultValue as T
            }
            throw InvalidBotPropertyValueException(value, descriptor)
        }

        val parsedValue: T
        try {
            parsedValue = descriptor.type.parse(value) as T
        } catch (e: Exception) {
            throw InvalidBotPropertyValueException(value, descriptor, e)
        }

        if (descriptor.acceptedValues.isEmpty() || descriptor.acceptedValues.contains(parsedValue)) {
            return parsedValue
        }
        throw InvalidBotPropertyValueException(value, descriptor)
    }

}
